clients/notification-pubsub/src/main/java/org/eclipse/hono/client/notification/pubsub/PubSubBasedNotificationSender.java

```diff
@@ -0,0 +1,78 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.client.notification.pubsub;
+
+import java.net.HttpURLConnection;
+import java.util.Objects;
+
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.client.pubsub.AbstractPubSubBasedMessageSender;
+import org.eclipse.hono.client.pubsub.PubSubMessageHelper;
+import org.eclipse.hono.client.pubsub.publisher.PubSubPublisherFactory;
+import org.eclipse.hono.notification.AbstractNotification;
+import org.eclipse.hono.notification.NotificationSender;
+
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+
+import io.opentracing.Tracer;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A client for publishing notifications to Google Pub/Sub.
+ */
+public class PubSubBasedNotificationSender extends AbstractPubSubBasedMessageSender implements NotificationSender {
+
+    private static final String TOPIC_ENDPOINT = "notification";
+
+    /**
+     * Creates a new PubSub-based notification sender.
+     *
+     * @param publisherFactory The factory to use for creating Pub/Sub publishers.
+     * @param projectId The identifier of the Google Cloud Project to connect to.
+     * @param tracer The OpenTracing tracer.
+     * @throws NullPointerException If any of the parameters are {@code null}.
+     */
+    public PubSubBasedNotificationSender(
+            final PubSubPublisherFactory publisherFactory,
+            final String projectId,
+            final Tracer tracer) {
+        super(publisherFactory, TOPIC_ENDPOINT, projectId, tracer);
+    }
+
+    @Override
+    public Future<Void> publish(final AbstractNotification notification) {
+        Objects.requireNonNull(notification);
+
+        if (!lifecycleStatus.isStarted()) {
+            return Future.failedFuture(
+                    new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "sender not started"));
+        }
+        final String topic = PubSubMessageHelper.getTopicName(TOPIC_ENDPOINT, notification.getType().getAddress());
+        final JsonObject value = JsonObject.mapFrom(notification);
+        final ByteString data = ByteString.copyFrom(value.toBuffer().getBytes());
+
+        final PubsubMessage pubsubMessage = PubsubMessage.newBuilder()
+                .setOrderingKey(notification.getType().getAddress())
+                .setData(data)
+                .build();
+        log.debug("sending notification to Pub/Sub [topic: {}, key: {}]", topic, notification.getKey());
+        return getOrCreatePublisher(topic).publish(pubsubMessage)
+                .onFailure(t -> {
+                    log.error("error publishing notification to Pub/Sub for notification [{}]", notification, t);
+                    throw new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR, t);
```

Comment:
IMHO this is not an internal error because the problem does not lie within Hono. FMPOV a Service Not Available would be a better match ...